main/src/ca/uwaterloo/flix/language/phase/Eraser.scala

```diff
@@ -213,69 +215,98 @@ object Eraser extends Phase[FinalAst.Root, FinalAst.Root]  {
-        ErasedAst.Term.Body.Lit(sym, tpe, loc)
+        ErasedAst.Term.Body.Lit(sym, visitTpe(tpe), loc)
     }
 
     /**
      * Translates the given `lattice0` to the ErasedAst.
      */
     def visitLatticeOps(lattice0: FinalAst.LatticeOps): ErasedAst.LatticeOps = lattice0 match {
       case FinalAst.LatticeOps(tpe, bot, equ, leq, lub, glb) =>
-        ErasedAst.LatticeOps(tpe, bot, equ, leq, lub, glb)
+        ErasedAst.LatticeOps(visitTpe(tpe), bot, equ, leq, lub, glb)
     }
 
     /**
      * Translates the given attribute `a` to the ErasedAst.
      */
     def visitAttribute(a: FinalAst.Attribute): ErasedAst.Attribute =
-      ErasedAst.Attribute(a.name, a.tpe)
+      ErasedAst.Attribute(a.name, visitTpe(a.tpe))
 
     /**
      * Translates the given formal param `p` to the ErasedAst.
      */
     def visitFormalParam(p: FinalAst.FormalParam): ErasedAst.FormalParam =
-      ErasedAst.FormalParam(p.sym, p.tpe)
+      ErasedAst.FormalParam(p.sym, visitTpe(p.tpe))
 
     /**
      * Translates the property `p` to the ErasedAst.
      */
     def visitProperty(p: FinalAst.Property): ErasedAst.Property =
       ErasedAst.Property(p.law, p.defn, visitExp(p.exp))
 
+    def visitTpe[T <: JType](tpe: MonoType): ErasedType[T] = (tpe match {
+      case MonoType.Unit => ErasedType.Unit()
+      case MonoType.Bool => ErasedType.Bool()
+      case MonoType.Char => ErasedType.Char()
+      case MonoType.Float32 => ErasedType.Float32()
+      case MonoType.Float64 => ErasedType.Float64()
+      case MonoType.Int8 => ErasedType.Int8()
+      case MonoType.Int16 => ErasedType.Int16()
+      case MonoType.Int32 => ErasedType.Int32()
+      case MonoType.Int64 => ErasedType.Int64()
+      case MonoType.BigInt => ErasedType.BigInt()
+      case MonoType.Str => ErasedType.Str()
+      case MonoType.Array(tpe) => ErasedType.Array(visitTpe(tpe))
+      case MonoType.Channel(tpe) => ErasedType.Channel(visitTpe(tpe))
+      case MonoType.Lazy(tpe) => ErasedType.Lazy(visitTpe(tpe))
+      case MonoType.Ref(tpe) => ErasedType.Ref(visitTpe(tpe))
+      case MonoType.Tuple(elms) => ErasedType.Tuple(elms.map(visitTpe))
+      case MonoType.Enum(sym, args) => ErasedType.Enum(sym, args.map(visitTpe))
+      case MonoType.Arrow(args, result) => ErasedType.Arrow(args.map(visitTpe), visitTpe(result))
+      case MonoType.RecordEmpty() => ErasedType.RecordEmpty()
+      case MonoType.RecordExtend(field, value, rest) => ErasedType.RecordExtend(field, visitTpe(value), visitTpe(rest))
+      case MonoType.SchemaEmpty() => ErasedType.SchemaEmpty()
+      case MonoType.SchemaExtend(name, tpe, rest) => ErasedType.SchemaExtend(name, visitTpe(tpe), visitTpe(rest))
+      case MonoType.Relation(tpes) => ErasedType.Relation(tpes.map(visitTpe))
+      case MonoType.Lattice(tpes) => ErasedType.Lattice(tpes.map(visitTpe))
+      case MonoType.Native(clazz) => ErasedType.Native(clazz)
+      case MonoType.Var(id) => ErasedType.Var(id)
+    }).asInstanceOf[ErasedType[T]]
+
     //
```

Comment:
It would probably be a good idea to move all these methods out into private methods not nested inside run.